File: src/api/agent_definitions.py
"""
子智能体定义管理 API

提供 /api/admin/agent-definitions 路由，管理子智能体定义和 system_prompt。
"""

import json
from datetime import datetime, date
from decimal import Decimal
from uuid import UUID
from typing import Optional, Dict, Any, List

from fastapi import APIRouter, HTTPException, Request
from fastapi.responses import JSONResponse
from loguru import logger
from pydantic import BaseModel, Field

from src.api.auth import get_current_user
from src.config.settings import settings
from src.services.subagent_definition_service import SubagentDefinitionService
from src.prompts.prompt_registry_service import PromptRegistryService


# ============== 请求模型 ==============

class CreateDefinitionRequest(BaseModel):
    agent_id: str = Field(..., min_length=1, max_length=100, description="智能体唯一标识")
    name: str = Field(..., min_length=1, max_length=100, description="显示名称")
    description: Optional[str] = None
    system_prompt: str = Field(..., min_length=1, description="System Prompt 内容")
    version: str = "1.0.0"
    author: Optional[str] = None
    triggers: Optional[Dict[str, Any]] = None
    tools: Optional[Dict[str, Any]] = None
    skills: Optional[Dict[str, Any]] = None
    context: Optional[Dict[str, Any]] = None
    delegatable_to: Optional[List[str]] = None
    allow_delegation: bool = True
    llm_provider: Optional[str] = None
    llm_model_codes: Optional[Dict[str, str]] = None
    reply_style: Optional[str] = None
    business_pages: Optional[List[Dict[str, Any]]] = None
    knowledge_sources: Optional[List[Dict[str, str]]] = None
    recap: Optional[Dict[str, Any]] = None
    chat_toolbar: Optional[List[str]] = None
    upload_accept: Optional[str] = None
    commit_message: str = "初始版本"


class UpdateDefinitionRequest(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    version: Optional[str] = None
    author: Optional[str] = None
    triggers: Optional[Dict[str, Any]] = None
    tools: Optional[Dict[str, Any]] = None
    skills: Optional[Dict[str, Any]] = None
    context: Optional[Dict[str, Any]] = None
    delegatable_to: Optional[List[str]] = None
    allow_delegation: Optional[bool] = None
    llm_provider: Optional[str] = None
    llm_model_codes: Optional[Dict[str, str]] = None
    reply_style: Optional[str] = None
    business_pages: Optional[List[Dict[str, Any]]] = None
    knowledge_sources: Optional[List[Dict[str, str]]] = None
    recap: Optional[Dict[str, Any]] = None
    chat_toolbar: Optional[List[str]] = None
    upload_accept: Optional[str] = None
    status: Optional[str] = None


class UpdateSystemPromptRequest(BaseModel):
    content: str = Field(..., min_length=1)
    commit_message: Optional[str] = None


# ============== 工具函数 ==============

def _serialize(obj):
    """递归将不可 JSON 序列化的类型转为可序列化的值"""
    if isinstance(obj, datetime):
        return obj.isoformat()
    if isinstance(obj, date):
        return obj.isoformat()
    if isinstance(obj, Decimal):
        return float(obj)
    if isinstance(obj, UUID):
        return str(obj)
    if isinstance(obj, dict):
        return {k: _serialize(v) for k, v in obj.items()}
    if isinstance(obj, list):
        return [_serialize(v) for v in obj]
    return obj


def _success(data=None, **kwargs) -> JSONResponse:
    body = {"success": True}
    if data is not None:
        body["data"] = _serialize(data)
    body.update(kwargs)
    return JSONResponse(body)


def _error(error: str, status_code: int = 400) -> JSONResponse:
    return JSONResponse({"success": False, "error": error}, status_code=status_code)


def _require_admin(request: Request):
    """平台管理员权限校验"""
    auth_header = request.headers.get("Authorization", "")
    logger.debug(f"agent-definitions auth: Authorization={auth_header[:20] + '...' if auth_header else 'MISSING'}")
    user = get_current_user(request)
    if not user:
        raise HTTPException(status_code=401, detail="未登录")

    admin_phones = getattr(settings, "admin", None)
    if admin_phones:
        phone_list = getattr(admin_phones, "phones", [])
        if user.get("phone", "") in phone_list:
            return user

    if user.get("role") == "platform_admin":
        return user

    raise HTTPException(status_code=403, detail="无管理员权限")


# ============== API 路由 ==============

router = APIRouter(prefix="/api/admin/agent-definitions", tags=["智能体定义管理"])


@router.get("")
async def list_definitions(
    request: Request,
    status: Optional[str] = None,
    page: int = 1,
    page_size: int = 20,
):
    """列出子智能体定义"""
    try:
        _require_admin(request)
        result = SubagentDefinitionService.list_definitions(status, page, page_size)
        return _success(result)
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"列出子智能体定义失败: {e}")
        return _error("获取列表失败")


@router.post("")
async def create_definition(request: Request, body: CreateDefinitionRequest):
    """创建子智能体定义 + 初始 system_prompt"""
    try:
        admin = _require_admin(request)

        existing = SubagentDefinitionService.get_definition(body.agent_id)
        if existing:
            return _error(f"智能体 {body.agent_id} 已存在")

        result = SubagentDefinitionService.create_definition(
            agent_id=body.agent_id,
            name=body.name,
            system_prompt=body.system_prompt,
            description=body.description,
            version=body.version,
            author=body.author,
            triggers=body.triggers,
            tools=body.tools,
            skills=body.skills,
            context=body.context,
            delegatable_to=body.delegatable_to,
            allow_delegation=body.allow_delegation,
            llm_provider=body.llm_provider,
            llm_model_codes=body.llm_model_codes,
            reply_style=body.reply_style,
            business_pages=body.business_pages,
            knowledge_sources=body.knowledge_sources,
            recap=body.recap,
            chat_toolbar=body.chat_toolbar,
            upload_accept=body.upload_accept,
            created_by=admin.get("user_id"),
            commit_message=body.commit_message,
        )
        if not result:
            return _error("创建失败")
        return _success(result)
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"创建子智能体定义失败: {e}")
        return _error("创建失败")


@router.get("/meta/tools")
async def list_tools_meta(request: Request):
    """获取工具元数据列表（供前端选择器使用）"""
    _require_admin(request)
    from src.core.agent import master_agent
    tool_registry = master_agent.tool_registry
    if not tool_registry:
        return _success([])
    tools = []
    for name, tool in tool_registry._tools.items():
        tools.append({
            "id": name,
            "name": getattr(tool, 'display_name', None) or name,
            "description": getattr(tool, 'description', '') or '',
        })
    return _success(tools)


@router.get("/meta/skills")
async def list_skills_meta(request: Request):
    """获取技能元数据列表（供前端选择器使用）

    展示全部基础目录已加载 skill（未经主智能体白名单过滤），
    这样管理员在配置子智能体 skills.allowed 时能看到全部可选 skill。
    """
    _require_admin(request)
    from src.core.agent import master_agent
    skill_registry = master_agent.skill_registry
    if not skill_registry:
        return _success([])
    skills = []
    for name in skill_registry.list_all_loaded_skills():
        skill = skill_registry.get_all_skill(name)
        skills.append({
            "id": name,
            "name": name,
            "description": skill.description if skill else '',
        })
    return _success(skills)


@router.get("/meta/reply-styles")
async def list_reply_styles_meta(request: Request):
    """获取回复风格元数据列表（供前端选择器使用）

    返回系统内置风格 + 当前租户自定义风格，与回复风格管理页面保持一致。
    """
    admin = _require_admin(request)
    tenant_id = admin.get("tenant_id")
    styles = []

    # 1. 系统内置风格（从数据库获取真实的 name 和 description）
    from src.saas.db.reply_style_db import ReplyStyleDB
    system_styles = ReplyStyleDB.list_system_styles()
    seen_ids = set()
    for s in system_styles:
        seen_ids.add(s["style_id"])
        styles.append({
            "id": s["style_id"],
            "name": s["name"],
            "description": s.get("description") or "",
        })

    # 2. 当前租户自定义风格（如果有）
    if tenant_id:
        from src.saas.db.reply_style_db import SYSTEM_TENANT
        tenant_styles = ReplyStyleDB.list_active_by_tenant(tenant_id)
        for s in tenant_styles:
            if s["style_id"] not in seen_ids and s["tenant_id"] != SYSTEM_TENANT:
                seen_ids.add(s["style_id"])
                styles.append({
                    "id": s["style_id"],
                    "name": s["name"],
                    "description": s.get("description") or "",
                })

    return _success(styles)


@router.get("/meta/recap-tasks")
async def list_recap_tasks_meta(request: Request):
    """获取 recap 任务元数据（供前端 recap 任务编辑器下拉使用）

    任务名来自适配器注册表 RECAP_TASK_ADAPTERS，触发时机来自 runner.VALID_WHEN 白名单。
    """
    _require_admin(request)
    from src.services.recap.tasks import RECAP_TASK_ADAPTERS
    from src.services.recap.runner import VALID_WHEN

    tasks = []
    for name, adapter_cls in RECAP_TASK_ADAPTERS.items():
        tasks.append({
            "name": name,
            "description": (getattr(adapter_cls, "__doc__", "") or "").strip().splitlines()[0] if getattr(adapter_cls, "__doc__", "") else "",
        })
    return _success({"tasks": tasks, "when_options": list(VALID_WHEN)})


# ============== Page Metadata ==============

from src.api.page_metadata import list_pages as _list_pages_meta, recommend_pages as _recommend_pages_meta


class RecommendPagesRequest(BaseModel):
    agent_description: str
    current_pages: List[str] = Field(default_factory=list)


@router.get("/meta/pages")
async def get_pages_meta(request: Request, domain: Optional[str] = None):
    """获取已发布的业务页面元数据列表（供前端选择器使用）"""
    return await _list_pages_meta(request, domain)


@router.post("/meta/pages/recommend")
async def post_pages_recommend(request: Request, body: RecommendPagesRequest):
    """AI 推荐与智能体描述相关的业务页面"""
    return await _recommend_pages_meta(request, body.model_dump())


@router.get("/{agent_id}")
async def get_definition(request: Request, agent_id: str):
    """获取子智能体定义详情"""
    try:
        _require_admin(request)
        result = SubagentDefinitionService.get_definition(agent_id)
        if not result:
            return _error(f"智能体 {agent_id} 不存在", 404)
        return _success(result)
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取子智能体定义失败: {e}")
        return _error("获取失败")


@router.put("/{agent_id}")
async def update_definition(request: Request, agent_id: str, body: UpdateDefinitionRequest):
    """更新子智能体定义（不含 system_prompt）

    使用 exclude_unset 而非 exclude_none，以便区分「未传入字段」和「显式传 null 清空」。
    这对 llm_provider / llm_model_codes 至关重要：用户需要能通过传 null 清空已有 LLM 配置。
    """
    try:
        admin = _require_admin(request)
        updates = body.model_dump(exclude_unset=True)
        if not updates:
            return _error("没有需要更新的字段")

        result = SubagentDefinitionService.update_definition(
            agent_id=agent_id,
            created_by=admin.get("user_id"),
            **updates,
        )
        if not result:
            return _error(f"智能体 {agent_id} 不存在或更新失败")

        # 刷新 master_agent 的 subagent_registry 内存配置，使 LLM/工具/技能等变更即时生效（无需重启容器）
        from src.core.agent import get_master_agent
        registry = get_master_agent().subagent_registry
        if registry:
            registry.load_from_db()

        return _success({"agent_id": agent_id})
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"更新子智能体定义失败: {e}")
        return _error("更新失败")


@router.delete("/{agent_id}")
async def delete_definition(request: Request, agent_id: str):
    """删除子智能体定义 + 关联的 Prompt"""
    try:
        _require_admin(request)
        result = SubagentDefinitionService.delete_definition(agent_id)
        if not result:
            return _error(f"智能体 {agent_id} 不存在或删除失败")
        return _success({"agent_id": agent_id})
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"删除子智能体定义失败: {e}")
        return _error("删除失败")


@router.put("/{agent_id}/system-prompt")
async def update_system_prompt(request: Request, agent_id: str, body: UpdateSystemPromptRequest):
    """提交 system_prompt 新版本并标记 production"""
    try:
        admin = _require_admin(request)
        result = SubagentDefinitionService.update_system_prompt(
            agent_id=agent_id,
            content=body.content,
            commit_message=body.commit_message,
            created_by=admin.get("user_id"),
        )
        if not result:
            return _error(f"智能体 {agent_id} 的 system_prompt 更新失败")
        return _success(result)
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"更新 system_prompt 失败: {e}")
        return _error("更新失败")


# ========== Prompt 版本管理（委托给 PromptRegistryService） ==========

def _get_prompt_id(agent_id: str) -> Optional[str]:
    prompt = PromptRegistryService.get_prompt_by_scope(None, "subagent", agent_id)
    return str(prompt["id"]) if prompt else None


@router.get("/{agent_id}/versions")
async def list_versions(request: Request, agent_id: str, page: int = 1, page_size: int = 20):
    """列出 system_prompt 版本"""
    try:
        _require_admin(request)
        prompt_id = _get_prompt_id(agent_id)
        if not prompt_id:
            return _error(f"智能体 {agent_id} 无关联 Prompt", 404)
        result = PromptRegistryService.list_versions(prompt_id, page, page_size)
        return _success(result)
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"列出版本失败: {e}")
        return _error("获取版本列表失败")


@router.get("/{agent_id}/versions/{version}")
async def get_version(request: Request, agent_id: str, version: int):
    """获取特定版本"""
    try:
        _require_admin(request)
        prompt_id = _get_prompt_id(agent_id)
        if not prompt_id:
            return _error(f"智能体 {agent_id} 无关联 Prompt", 404)
        result = PromptRegistryService.get_version(prompt_id, version)
        if not result:
            return _error(f"版本 {version} 不存在", 404)
        return _success(result)
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取版本失败: {e}")
        return _error("获取版本失败")


@router.get("/{agent_id}/versions/diff")
async def diff_versions(request: Request, agent_id: str, v1: int, v2: int):
    """两个版本 diff 对比"""
    try:
        _require_admin(request)
        prompt_id = _get_prompt_id(agent_id)
        if not prompt_id:
            return _error(f"智能体 {agent_id} 无关联 Prompt", 404)
        result = PromptRegistryService.diff_versions(prompt_id, v1, v2)
        if not result:
            return _error("版本不存在", 404)
        return _success(result)
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"版本对比失败: {e}")
        return _error("版本对比失败")


@router.get("/{agent_id}/draft")
async def get_draft(request: Request, agent_id: str):
    """获取草稿"""
    try:
        _require_admin(request)
        prompt_id = _get_prompt_id(agent_id)
        if not prompt_id:
            return _error(f"智能体 {agent_id} 无关联 Prompt", 404)
        result = PromptRegistryService.get_draft(prompt_id)
        return _success(result)
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取草稿失败: {e}")
        return _error("获取草稿失败")


@router.put("/{agent_id}/draft")
async def save_draft(request: Request, agent_id: str, body: dict):
    """保存草稿"""
    try:
        admin = _require_admin(request)
        prompt_id = _get_prompt_id(agent_id)
        if not prompt_id:
            return _error(f"智能体 {agent_id} 无关联 Prompt", 404)
        result = PromptRegistryService.save_draft(
            prompt_id=prompt_id,
            content=body.get("content", ""),
            variables=body.get("variables"),
            base_version=body.get("base_version"),
            updated_by=admin.get("user_id"),
        )
        return _success(result)
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"保存草稿失败: {e}")
        return _error("保存草稿失败")


@router.delete("/{agent_id}/draft")
async def delete_draft(request: Request, agent_id: str):
    """删除草稿"""
    try:
        _require_admin(request)
        prompt_id = _get_prompt_id(agent_id)
        if not prompt_id:
            return _error(f"智能体 {agent_id} 无关联 Prompt", 404)
        PromptRegistryService.delete_draft(prompt_id)
        return _success()
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"删除草稿失败: {e}")
        return _error("删除草稿失败")


@router.post("/{agent_id}/draft/commit")
async def commit_draft(request: Request, agent_id: str, body: dict):
    """提交草稿为新版本"""
    try:
        admin = _require_admin(request)
        prompt_id = _get_prompt_id(agent_id)
        if not prompt_id:
            return _error(f"智能体 {agent_id} 无关联 Prompt", 404)
        result = PromptRegistryService.commit_draft(
            prompt_id=prompt_id,
            commit_message=body.get("commit_message"),
            created_by=admin.get("user_id"),
        )
        if not result or not result.get("version"):
            return _error("提交草稿失败")
        return _success(result)
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"提交草稿失败: {e}")
        return _error("提交草稿失败")


@router.get("/{agent_id}/labels")
async def list_labels(request: Request, agent_id: str):
    """列出标签"""
    try:
        _require_admin(request)
        prompt_id = _get_prompt_id(agent_id)
        if not prompt_id:
            return _error(f"智能体 {agent_id} 无关联 Prompt", 404)
        result = PromptRegistryService.list_labels(prompt_id)
        return _success(result)
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"列出标签失败: {e}")
        return _error("获取标签失败")


@router.put("/{agent_id}/labels/{label}")
async def set_label(request: Request, agent_id: str, label: str, body: dict):
    """设置标签"""
    try:
        admin = _require_admin(request)
        prompt_id = _get_prompt_id(agent_id)
        if not prompt_id:
            return _error(f"智能体 {agent_id} 无关联 Prompt", 404)
        version = body.get("version")
        if not version:
            return _error("缺少 version 参数")
        result = PromptRegistryService.set_label(
            prompt_id=prompt_id,
            label=label,
            version=version,
            created_by=admin.get("user_id"),
        )
        if not result:
            return _error("设置标签失败")
        return _success(result)
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"设置标签失败: {e}")
        return _error("设置标签失败")


@router.delete("/{agent_id}/labels/{label}")
async def delete_label(request: Request, agent_id: str, label: str):
    """删除标签"""
    try:
        _require_admin(request)
        prompt_id = _get_prompt_id(agent_id)
        if not prompt_id:
            return _error(f"智能体 {agent_id} 无关联 Prompt", 404)
        result = PromptRegistryService.delete_label(prompt_id, label)
        if not result:
            return _error("删除标签失败")
        return _success()
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"删除标签失败: {e}")
        return _error("删除标签失败")

File: frontend/src/api/agentDefinitions.ts
/**
 * 子智能体定义管理 API
 */
const API_BASE = `${import.meta.env.VITE_API_BASE_URL || '/api'}/admin/agent-definitions`

function getAuthHeaders(): Record<string, string> {
  const token = localStorage.getItem('portal_token')
  return token ? { Authorization: `Bearer ${token}` } : {}
}

async function handleResponse<T>(response: Response): Promise<T> {
  const data = await response.json()
  if (!response.ok) {
    throw new Error(data.error || `请求失败: ${response.status}`)
  }
  return data
}

// ============== 类型定义 ==============

export interface AgentDefinition {
  id: string
  agent_id: string
  name: string
  description: string | null
  version: string
  author: string | null
  triggers: Record<string, any>
  tools: Record<string, any>
  skills: Record<string, any>
  context: Record<string, any>
  delegatable_to: string[]
  allow_delegation: boolean
  llm_provider: string | null
  reply_style: string | null
  business_pages: any[] | null
  knowledge_sources: { source_type: string; display_name: string }[]
  status: string
  created_by: string | null
  updated_by: string | null
  created_at: string
  updated_at: string
  // 关联的 prompt 信息
  prompt_id?: string
  latest_version?: number
  production_version?: number
}

export interface DefinitionListResult {
  total: number
  page: number
  page_size: number
  items: AgentDefinition[]
}

export interface PromptVersion {
  id: string
  prompt_id: string
  version: number
  content: string
  commit_message: string | null
  content_hash: string | null
  parent_version: number | null
  created_by: string | null
  created_at: string
}

export interface PromptLabel {
  label: string
  version_id: string
  version: number | null
  created_by: string | null
  updated_by: string | null
  created_at: string
  updated_at: string
}

export interface DiffResult {
  from: { version: number; content: string; created_at: string; created_by: string | null }
  to: { version: number; content: string; created_at: string; created_by: string | null }
}

// ============== 定义 CRUD ==============

export async function listDefinitions(params?: {
  status?: string; page?: number; page_size?: number
}): Promise<{ success: boolean; data: DefinitionListResult }> {
  const query = new URLSearchParams()
  if (params?.status) query.set('status', params.status)
  if (params?.page) query.set('page', String(params.page))
  if (params?.page_size) query.set('page_size', String(params.page_size))
  const qs = query.toString()
  const headers = getAuthHeaders()
  const response = await fetch(`${API_BASE}${qs ? '?' + qs : ''}`, { headers })
  return handleResponse(response)
}

export async function getDefinition(agentId: string): Promise<{ success: boolean; data: AgentDefinition }> {
  const response = await fetch(`${API_BASE}/${encodeURIComponent(agentId)}`, { headers: getAuthHeaders() })
  return handleResponse(response)
}

export async function createDefinition(data: {
  agent_id: string
  name: string
  system_prompt: string
  description?: string
  tools?: Record<string, any>
  skills?: Record<string, any>
  context?: Record<string, any>
  llm_provider?: string
  reply_style?: string
}): Promise<{ success: boolean; data: any }> {
  const response = await fetch(`${API_BASE}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...getAuthHeaders() },
    body: JSON.stringify(data),
  })
  return handleResponse(response)
}

export async function updateDefinition(agentId: string, data: Record<string, any>): Promise<{ success: boolean; data: any }> {
  const response = await fetch(`${API_BASE}/${encodeURIComponent(agentId)}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json', ...getAuthHeaders() },
    body: JSON.stringify(data),
  })
  return handleResponse(response)
}

export async function deleteDefinition(agentId: string): Promise<{ success: boolean }> {
  const response = await fetch(`${API_BASE}/${encodeURIComponent(agentId)}`, {
    method: 'DELETE',
    headers: getAuthHeaders(),
  })
  return handleResponse(response)
}

// ============== System Prompt 管理 ==============

export async function updateSystemPrompt(agentId: string, data: {
  content: string; commit_message?: string
}): Promise<{ success: boolean; data: any }> {
  const response = await fetch(`${API_BASE}/${encodeURIComponent(agentId)}/system-prompt`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json', ...getAuthHeaders() },
    body: JSON.stringify(data),
  })
  return handleResponse(response)
}

export async function listVersions(agentId: string, page = 1, pageSize = 20): Promise<{ success: boolean; data: { total: number; page: number; page_size: number; items: PromptVersion[] } }> {
  const response = await fetch(`${API_BASE}/${encodeURIComponent(agentId)}/versions?page=${page}&page_size=${pageSize}`, { headers: getAuthHeaders() })
  return handleResponse(response)
}

export async function getVersion(agentId: string, version: number): Promise<{ success: boolean; data: PromptVersion }> {
  const response = await fetch(`${API_BASE}/${encodeURIComponent(agentId)}/versions/${version}`, { headers: getAuthHeaders() })
  return handleResponse(response)
}

export async function diffVersions(agentId: string, v1: number, v2: number): Promise<{ success: boolean; data: DiffResult }> {
  const response = await fetch(`${API_BASE}/${encodeURIComponent(agentId)}/versions/diff?v1=${v1}&v2=${v2}`, { headers: getAuthHeaders() })
  return handleResponse(response)
}

export async function getDraft(agentId: string): Promise<{ success: boolean; data: any }> {
  const response = await fetch(`${API_BASE}/${encodeURIComponent(agentId)}/draft`, { headers: getAuthHeaders() })
  return handleResponse(response)
}

export async function saveDraft(agentId: string, data: {
  content: string; base_version?: number
}): Promise<{ success: boolean; data: any }> {
  const response = await fetch(`${API_BASE}/${encodeURIComponent(agentId)}/draft`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json', ...getAuthHeaders() },
    body: JSON.stringify(data),
  })
  return handleResponse(response)
}

export async function commitDraft(agentId: string, data: {
  commit_message?: string
}): Promise<{ success: boolean; data: any }> {
  const response = await fetch(`${API_BASE}/${encodeURIComponent(agentId)}/draft/commit`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...getAuthHeaders() },
    body: JSON.stringify(data),
  })
  return handleResponse(response)
}

export async function listLabels(agentId: string): Promise<{ success: boolean; data: PromptLabel[] }> {
  const response = await fetch(`${API_BASE}/${encodeURIComponent(agentId)}/labels`, { headers: getAuthHeaders() })
  return handleResponse(response)
}

export async function setLabel(agentId: string, label: string, version: number): Promise<{ success: boolean; data: PromptLabel }> {
  const response = await fetch(`${API_BASE}/${encodeURIComponent(agentId)}/labels/${encodeURIComponent(label)}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json', ...getAuthHeaders() },
    body: JSON.stringify({ version }),
  })
  return handleResponse(response)
}

// ============== Meta (Tools/Skills/ReplyStyles Pickers) ==============

export interface ToolMeta {
  id: string
  name: string
  description: string
}

export interface SkillMeta {
  id: string
  name: string
  description: string
}

export interface ReplyStyleMeta {
  id: string
  name: string
  description: string
}

export async function listToolsMeta(): Promise<{ success: boolean; data: ToolMeta[] }> {
  const response = await fetch(`${API_BASE}/meta/tools`, { headers: getAuthHeaders() })
  return handleResponse(response)
}

export async function listSkillsMeta(): Promise<{ success: boolean; data: SkillMeta[] }> {
  const response = await fetch(`${API_BASE}/meta/skills`, { headers: getAuthHeaders() })
  return handleResponse(response)
}

export async function listReplyStylesMeta(): Promise<{ success: boolean; data: ReplyStyleMeta[] }> {
  const response = await fetch(`${API_BASE}/meta/reply-styles`, { headers: getAuthHeaders() })
  return handleResponse(response)
}

// ============== Sections Management ==============

export interface PromptSection {
  agent_id: string
  section_key: string
  content: string
  updated_by: string | null
  created_at?: string
  updated_at?: string
}

export async function getSectionKeys(agentId: string): Promise<{ success: boolean; data: string[] }> {
  const response = await fetch(`${API_BASE}/${encodeURIComponent(agentId)}/sections/keys`, { headers: getAuthHeaders() })
  return handleResponse(response)
}

export async function getSections(agentId: string): Promise<{ success: boolean; data: PromptSection[] }> {
  const response = await fetch(`${API_BASE}/${encodeURIComponent(agentId)}/sections`, { headers: getAuthHeaders() })
  return handleResponse(response)
}

export async function saveSection(
  agentId: string,
  sectionKey: string,
  content: string,
): Promise<{ success: boolean; data: PromptSection }> {
  const response = await fetch(`${API_BASE}/${encodeURIComponent(agentId)}/sections/${encodeURIComponent(sectionKey)}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json', ...getAuthHeaders() },
    body: JSON.stringify({ content, section_key: sectionKey }),
  })
  return handleResponse(response)
}

export async function optimizeSection(
  agentId: string,
  sectionKey: string,
  data: { content: string; agent_name?: string; agent_description?: string },
): Promise<{ success: boolean; data: { content: string } }> {
  const response = await fetch(`${API_BASE}/${encodeURIComponent(agentId)}/sections/${encodeURIComponent(sectionKey)}/optimize`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...getAuthHeaders() },
    body: JSON.stringify({ ...data, section_key: sectionKey }),
  })
  return handleResponse(response)
}
